File: mcmc/MCMC.py
import cPickle as pickle
import numpy as np
import matplotlib.pyplot as plt
import os
import gzip
from threading import Thread
from multiprocessing import Process
import pandas as pd
import time

mean=[1,400,30,50]

def default_logp(value):
    log=0
    for i in range(4):
        var=-((value[i]-mean[i])/mean[i])**2
        log+=var
        return log
    
class MCMC(Process):
    chunkSize=1e6
    
    def __init__(self,filename,initialCondition,realValues=[]):
        Process.__init__(self)
        self.realValues=realValues
        self.filename=filename
        self.nStep=10
        self.current_point=initialCondition # the chain starting point
        self.nVar=len(initialCondition)
        self.verbose=False
        self.trace=np.zeros((self.chunkSize,self.nVar))
        self.log_likelihood=np.zeros(self.chunkSize) # the chain containing the log likelihood of all accepted points
        self.sigma=[1 for i in range(self.nVar)] # the sigmas for the proposal function
        self.logp=default_logp
        self.chunkStepNumber=0
        self.seed=0

    def run(self):
        self.loop()

    def saveMetaData(self):
        print 'save metadata'
        f=open(self.filename,'wb')
        pickle.dump(self.metadata,f,2)
        f.close()

    def saveChunk(self):
        print 'saving chunk...'
        #f=gzip.open(self.filename,'a')
        df=pd.DataFrame(self.trace[:self.chunkStepNumber])
        dfLog=pd.DataFrame(self.log_likelihood[:self.chunkStepNumber],columns=['log'])
        dfTotal=df.join(dfLog)
        print 'df done'
        f=open(self.filename,'ab')
        print 'file opened'

        pickle.dump(dfTotal, f, protocol=2)
          
        f.close()
        print 'file closed'
        self.log_likelihood=np.zeros(self.chunkSize)
        self.trace=np.zeros((self.chunkSize,self.nVar))
        self.chunkStepNumber=0
        print 'chunk saved'

    def setSigmas(self, sigma):
        self.sigma=sigma

    def proposal_function(self,previous_point):
        return previous_point+self.sigma*np.random.standard_normal(self.nVar)

    def setProposalFunction(self,f):
        self.proposal_function=f

    def setLogLikelihoodFunction(self,f):
        self.logp=f

    def setSteps(self,steps):
        self.nStep=steps

    def setVerbose(self,isVerbose):
        self.verbose=isVerbose

    def updateStep(self):
        if self.verbose: print 'accepted'
        self.current_point=self.proposed_point
        self.current_log_likelihood=self.proposed_log_likelihood
        self.trace[self.chunkStepNumber]=self.current_point
        self.log_likelihood[self.chunkStepNumber]=self.current_log_likelihood
        self.chunkStepNumber+=1

    def loop(self):
        np.random.seed()

        self.metadata={'initialPoints':self.current_point,'nStep':self.nStep,'sigmas':self.sigma,'nVar':self.nVar,'realValues':self.realValues}
        self.saveMetaData()

        self.current_log_likelihood=self.logp(self.current_point)

        for i in range(self.nStep):
            if i%100000 == 0: print '{}/{} : {}%'.format(i,self.nStep,int(float(i)/float(self.nStep)*100))
            self.proposed_point=self.proposal_function(self.current_point)
            self.proposed_log_likelihood=self.logp(self.proposed_point)
            
            self.the_likelihood_ratio=np.exp(self.proposed_log_likelihood-self.current_log_likelihood)

            if self.verbose:
                print 'current: {}'.format(self.current_point)
                print 'proposed: {}'.format(self.proposed_point)
                
            if self.the_likelihood_ratio > 1:
                self.updateStep()
                
            else:
                if np.random.rand() < self.the_likelihood_ratio:
                    self.updateStep()

            if self.verbose:
                print 'current_log_likelihood: {}'.format(self.current_log_likelihood)    
                print 'proposed_log_likelihood: {}'.format(self.proposed_log_likelihood)
                print 'likelihood_ratio: {}'.format(self.the_likelihood_ratio)

            if self.chunkStepNumber >= self.chunkSize:
                self.saveChunk()


        print 'step number'+str(self.chunkStepNumber)
        if self.chunkStepNumber > 0: self.saveChunk()

File: mcmc/mcmc.cpp
#include "mcmc.hpp"

template <class ProposalFunction > MCMC<ProposalFunction >::MCMC(std::string name, ModelBase & _model):
    model(_model),
    realValues(model.GetRealValues()),
    current_point(model.GetInitialConditions()),
    initialConditions(model.GetInitialConditions()),
    proposalFunction(model.GetInitialConditions()),
    nVar(initialConditions.size()),
    nThreads(1),
    chunkSize(maxRAM / ( (nVar+1)*sizeof(float)) / nThreads),
    spectators(),
    log_likelihood(chunkSize),
    trace(nVar),
    filename(name),
    current_log_likelihood(0),
    seed(0),
    chunkStepNumber(0),
    nStep(100),
    verbose(false)
{

    if( mkdir(filename.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) != 0){
        int i = 0;
        while( mkdir( (filename+generalUtils::toString(i)).c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) != 0 ){
            i++;
        }
        filename = filename+generalUtils::toString(i);
    }

    for(int i = 0;i<nVar;i++) trace[i] = std::vector<float>(chunkSize);

    // construct a trivial random generator engine from a time-based seed:
    seed = std::chrono::system_clock::now().time_since_epoch().count();
    generator.seed(seed);
}

template <class ProposalFunction > void MCMC<ProposalFunction >::loop(){
    std::cout << "nStep : " << nStep/1e6 << " millions" << std::endl;
    std::cout << "chunkSize : " << chunkSize/1e6 << " MBytes" << std::endl;
    std::cout << "maxRAM : " << maxRAM/1e6 << " MBytes" << std::endl;
    std::cout << "nVar : " << nVar << std::endl;
    std::cout << "sizeof(float) : " << sizeof(float) << std::endl;

    sleep(1);

    saveMetaData();

    current_log_likelihood = model.GetLogLikelihood(initialConditions);
        
    for( int i = 0; i < nStep; i++)
        {
            if( i%30000 == 0) printf("%i/%i : %i%%\n",i, nStep, int(float(i)/float(nStep)*100));

            auto proposed_point = proposalFunction.proposePoint(current_point);

            float proposed_log_likelihood = model.GetLogLikelihood(proposed_point);

            float the_likelihood_ratio = exp(proposed_log_likelihood-current_log_likelihood);

            if( verbose ){

                std::cout << "fluxP "; for(auto v: proposed_point.fluxP) std::cout << v << " "; std::cout << "\n";
                std::cout << "fluxD "; for(auto v: proposed_point.fluxD) std::cout << v << " "; std::cout << "\n";

                std::cout << "current_log_likelihood : " << current_log_likelihood << std::endl;
                std::cout << "proposed_log_likelihood : " << proposed_log_likelihood << std::endl;
                std::cout << "the_likelihood_ratio : " << the_likelihood_ratio << std::endl;
            }

            // Accept the proposed point if:
            // the likelihood of the proposed point is bigger than previous likelihood
            // OR
            // if smaller, accept it with a chance equal to the likelihood ratio
            // rand() / (float) RAND_MAX, generate a random number between 0 and 1
            if( the_likelihood_ratio > 1 || rand() / (float) RAND_MAX < the_likelihood_ratio ){
                updateStep(proposed_point, proposed_log_likelihood);
            } else addCurrentPointToChain();

            if( chunkStepNumber >= chunkSize) saveChunk();

        }
        
    std::cout << "chunkStepNumber : " << chunkStepNumber << std::endl;
    if( chunkStepNumber > 0 ) saveChunk();
}

template <class ProposalFunction > void MCMC<ProposalFunction>::setSpectator(std::function<float(const ModelBase&)> f){
    spectators.push_back(f);
}

int main(int argc, char** argv){
    int c;
    int nStep = 0;
    std::string name = "test";
    std::string maskFile = ""; 
    bool verbose = false;
    float alphaRegularization = 0;
    while((c =  getopt(argc, argv, "n:f:v:a:m:")) != EOF)
        {
            switch (c)
                {
                case 'n':
                    nStep = generalUtils::stringTo<int>(optarg);
                    break;
                case 'f':
                    name = optarg;
                    break;
                case 'v':
                    verbose = true;
                    break;
                case 'a':
                    alphaRegularization = generalUtils::stringTo<float>(optarg);
                    break;
                case 'm':
                    maskFile = optarg;
                    break;
                default:
                    break;
                }
        }

    std::clock_t start = std::clock();

    //    MCMC<RealisticToyModel, ProposalFunction > a(name);

    
    RealDataModel model;
    if(maskFile != "") model.SetMask(maskFile);
    model.setRegularizationFactor(alphaRegularization);

    MCMC<ProposalFunction > a(name,model);
    a.setVerbose(verbose);
    //    a.setSpectator( &RealDataModel::regularizationTermAccessor );

    if( nStep > 0 ) a.setSteps(nStep);
    a.run();
    std::cout << "sizeof(a) : " << sizeof(a) << std::endl;
    std::cout << "Time : " << (std::clock() - start) / (float)(CLOCKS_PER_SEC) << " s" << std::endl;
    std::cout << "Processing of : " << name << " done !" << std::endl;
    return 0;
}

